File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/RelatedBusinessUnit.cpp
// RelatedBusinessUnit.cpp 
#include "RelatedBusinessUnit.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

RelatedBusinessUnit::RelatedBusinessUnit(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //businessUnitReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessUnitReferenceNode = xmlNode->FirstChildElement("businessUnitReference");

   if(businessUnitReferenceNode){businessUnitReferenceIsNull_ = false;}
   else{businessUnitReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessUnitReferenceNode , address : " << businessUnitReferenceNode << std::endl;
   #endif
   if(businessUnitReferenceNode)
   {
      if(businessUnitReferenceNode->Attribute("href") || businessUnitReferenceNode->Attribute("id"))
      {
          if(businessUnitReferenceNode->Attribute("id"))
          {
             businessUnitReferenceIDRef_ = businessUnitReferenceNode->Attribute("id");
             businessUnitReference_ = boost::shared_ptr<BusinessUnitReference>(new BusinessUnitReference(businessUnitReferenceNode));
             businessUnitReference_->setID(businessUnitReferenceIDRef_);
             IDManager::instance().setID(businessUnitReferenceIDRef_,businessUnitReference_);
          }
          else if(businessUnitReferenceNode->Attribute("href")) { businessUnitReferenceIDRef_ = businessUnitReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessUnitReference_ = boost::shared_ptr<BusinessUnitReference>(new BusinessUnitReference(businessUnitReferenceNode));}
   }

   //roleNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* roleNode = xmlNode->FirstChildElement("role");

   if(roleNode){roleIsNull_ = false;}
   else{roleIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- roleNode , address : " << roleNode << std::endl;
   #endif
   if(roleNode)
   {
      if(roleNode->Attribute("href") || roleNode->Attribute("id"))
      {
          if(roleNode->Attribute("id"))
          {
             roleIDRef_ = roleNode->Attribute("id");
             role_ = boost::shared_ptr<BusinessUnitRole>(new BusinessUnitRole(roleNode));
             role_->setID(roleIDRef_);
             IDManager::instance().setID(roleIDRef_,role_);
          }
          else if(roleNode->Attribute("href")) { roleIDRef_ = roleNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { role_ = boost::shared_ptr<BusinessUnitRole>(new BusinessUnitRole(roleNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<BusinessUnitReference> RelatedBusinessUnit::getBusinessUnitReference()
{
   if(!this->businessUnitReferenceIsNull_){
        if(businessUnitReferenceIDRef_ != ""){
             return boost::shared_static_cast<BusinessUnitReference>(IDManager::instance().getID(businessUnitReferenceIDRef_));
        }else{
             return this->businessUnitReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessUnitReference>();
   }
}
boost::shared_ptr<BusinessUnitRole> RelatedBusinessUnit::getRole()
{
   if(!this->roleIsNull_){
        if(roleIDRef_ != ""){
             return boost::shared_static_cast<BusinessUnitRole>(IDManager::instance().getID(roleIDRef_));
        }else{
             return this->role_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessUnitRole>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/RelatedBusinessUnit.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class RelatedBusinessUnit : ISerialized
    {
        public RelatedBusinessUnit(XmlNode xmlNode)
        {
            XmlNode businessUnitReferenceNode = xmlNode.SelectSingleNode("businessUnitReference");
            
            if (businessUnitReferenceNode != null)
            {
                if (businessUnitReferenceNode.Attributes["href"] != null || businessUnitReferenceNode.Attributes["id"] != null) 
                {
                    if (businessUnitReferenceNode.Attributes["id"] != null) 
                    {
                        businessUnitReferenceIDRef_ = businessUnitReferenceNode.Attributes["id"].Value;
                        BusinessUnitReference ob = new BusinessUnitReference(businessUnitReferenceNode);
                        IDManager.SetID(businessUnitReferenceIDRef_, ob);
                    }
                    else if (businessUnitReferenceNode.Attributes["href"] != null)
                    {
                        businessUnitReferenceIDRef_ = businessUnitReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessUnitReference_ = new BusinessUnitReference(businessUnitReferenceNode);
                    }
                }
                else
                {
                    businessUnitReference_ = new BusinessUnitReference(businessUnitReferenceNode);
                }
            }
            
        
            XmlNode roleNode = xmlNode.SelectSingleNode("role");
            
            if (roleNode != null)
            {
                if (roleNode.Attributes["href"] != null || roleNode.Attributes["id"] != null) 
                {
                    if (roleNode.Attributes["id"] != null) 
                    {
                        roleIDRef_ = roleNode.Attributes["id"].Value;
                        BusinessUnitRole ob = new BusinessUnitRole(roleNode);
                        IDManager.SetID(roleIDRef_, ob);
                    }
                    else if (roleNode.Attributes["href"] != null)
                    {
                        roleIDRef_ = roleNode.Attributes["href"].Value;
                    }
                    else
                    {
                        role_ = new BusinessUnitRole(roleNode);
                    }
                }
                else
                {
                    role_ = new BusinessUnitRole(roleNode);
                }
            }
            
        
        }
        
    
        #region BusinessUnitReference_
        private BusinessUnitReference businessUnitReference_;
        public BusinessUnitReference BusinessUnitReference_
        {
            get
            {
                if (this.businessUnitReference_ != null)
                {
                    return this.businessUnitReference_; 
                }
                else if (this.businessUnitReferenceIDRef_ != null)
                {
                    businessUnitReference_ = IDManager.getID(businessUnitReferenceIDRef_) as BusinessUnitReference;
                    return this.businessUnitReference_; 
                }
                else
                {
                      return this.businessUnitReference_; 
                }
            }
            set
            {
                if (this.businessUnitReference_ != value)
                {
                    this.businessUnitReference_ = value;
                }
            }
        }
        #endregion
        
        public string businessUnitReferenceIDRef_ { get; set; }
        #region Role_
        private BusinessUnitRole role_;
        public BusinessUnitRole Role_
        {
            get
            {
                if (this.role_ != null)
                {
                    return this.role_; 
                }
                else if (this.roleIDRef_ != null)
                {
                    role_ = IDManager.getID(roleIDRef_) as BusinessUnitRole;
                    return this.role_; 
                }
                else
                {
                      return this.role_; 
                }
            }
            set
            {
                if (this.role_ != value)
                {
                    this.role_ = value;
                }
            }
        }
        #endregion
        
        public string roleIDRef_ { get; set; }
        
    
        
    
    }
    
}

